crates/engine/tree/src/engine.rs

```diff
@@ -0,0 +1,212 @@
+};
+use tokio::sync::mpsc;
+
+/// Advances the chain based on incoming requests.
+///
+/// This is a general purpose request handler with network access.
+/// This type listens for incoming messages and processes them via the configured request handler.
+///
+/// ## Overview
+///
+/// This type is an orchestrator for incoming messages and responsible for delegating requests
+/// received from the CL to the handler.
+///
+/// It is responsible for handling the following:
+/// - Downloading blocks on demand from the network if requested by the [`EngineApiRequestHandler`].
+///
+/// The core logic is part of the [`EngineRequestHandler`], which is responsible for processing the
+/// incoming requests.
+#[derive(Debug)]
+pub struct EngineHandler<T, S, D> {
+    /// Processes requests.
+    ///
+    /// This type is responsible for processing incoming requests.
+    handler: T,
+    /// Receiver for incoming requests that need to be processed.
+    incoming_requests: S,
+    /// A downloader to download blocks on demand.
+    downloader: D,
+}
+
+impl<T, S, D> EngineHandler<T, S, D> {
+    /// Creates a new [`EngineHandler`] with the given handler and downloader.
+    pub const fn new(handler: T, downloader: D, incoming_requests: S) -> Self
+    where
+        T: EngineRequestHandler,
+    {
+        Self { handler, incoming_requests, downloader }
+    }
+}
+
+impl<T, S, D> ChainHandler for EngineHandler<T, S, D>
+where
+    T: EngineRequestHandler,
+    S: Stream<Item = T::Request> + Send + Sync + Unpin + 'static,
+    D: BlockDownloader,
+{
+    type Event = T::Event;
+
+    fn on_event(&mut self, event: FromOrchestrator) {
+        // delegate event to the handler
+        self.handler.on_event(event.into());
+    }
+
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<HandlerEvent<Self::Event>> {
+        loop {
+            // drain the handler first
+            while let Poll::Ready(ev) = self.handler.poll(cx) {
+                match ev {
+                    RequestHandlerEvent::Idle => break,
+                    RequestHandlerEvent::HandlerEvent(ev) => {
+                        return match ev {
+                            HandlerEvent::Pipeline(target) => {
+                                // bubble up pipeline request
+                                self.downloader.on_action(DownloadAction::Clear);
+                                Poll::Ready(HandlerEvent::Pipeline(target))
+                            }
+                            HandlerEvent::Event(ev) => {
+                                // bubble up the event
+                                Poll::Ready(HandlerEvent::Event(ev))
+                            }
+                        }
+                    }
+                    RequestHandlerEvent::Download(req) => {
+                        // delegate download request to the downloader
+                        self.downloader.on_action(DownloadAction::Download(req));
+                    }
+                }
+            }
```

Comment:
can we use macro `poll_nested_stream_with_budget` to avoid starvation here. the macro can be moved from network crate to a common utility crate.

Reply:
out of scope for this pr